### ISSUE TITLE
[Proposal] make backup operator and restore operator controller like the operator

### ISSUE THREAD

**Issue by rjtsdl** [coreos/etcd-operator#1876]
Currently, etcd-operator is structed differently with backup and restore operator. 
**Etcd-operator**: Watch for etcdcluster CRD. And spin a control loop for each etcdCluster instance.
**Backup/Restore operators**: Watch for etcdbackup/etcdrestore CRD, Add the event to work queue. Process the work queue one by one.

The difference is, for each etcdCluster instance, there is a loop monitoring and reconciling for it. No loop for etcdbackup/etcdrestore instance.

We do have needs to have a control loop for every backup/restore instance. For example,
1. we need backup schedule loop up and running, to periodically take snapshot. #1841 
2. we need restore control loop up and running, to detect disaster, and trigger restore operation.

cc: @khenidak @shrutir25

**Comment by rjtsdl** [comment]
@hongchaodeng @xiang90 does it make sense?
I was about to address #1841. This is one thing I propose to have before addressing that.

**Comment by hongchaodeng** [comment]
@rjtsdl 
Can you submit a design proposal? See other examples in https://github.com/coreos/etcd-operator/tree/master/doc/design

**Comment by rjtsdl** [comment]
@hongchaodeng, it doesn't change the current design at all. Only that, every backup instance has its own control loop. 

It would be more straightforward, if i just draft a PR for it. On it.

**Comment by fanminshi** [comment]
@rjtsdl 
`we need backup schedule loop up and running, to periodically take snapshot.`
 the periodically backup is not really a reconciliation loop. it is just a feature/enhancement. not all operator needs an reconciliation loop.

`we need restore control loop up and running, to detect disaster, and trigger restore operation.`
This one is more like an reconciliation loop that etcd-operator has.  However, this is a bit tricky to get right. Who detects the failure of node? should etcd operator detect failure and instruct restore operator to restore it? or it is the job of restore operator to detect failure and recovery the node? as you can see. there will be some interaction between restore-operator and etcd-operator to get restore reconciliation loop right. The latter requires a design proposal since i don't think adding a  reconciliation loop in restore will be enough also requiring more thought.

**Comment by rjtsdl** [comment]
@fanminshi 

You are right, it could be a new feature. It is not necessary to be called reconcile loop. But we need a loop for each CRD instance.

I am currently working on a PR to make code struct in a fashion that, it is easy to add the function in the loop. I will make another PR to do the actual backup schedule feature.

For restore-operator, there are questions like you raised. I am currently only do it for backup. 
Can you elaborate more on `i don't think adding a reconciliation loop in restore will be enough` ? That could be in the design proposal.

**Comment by fanminshi** [comment]
@rjtsdl `i don't think adding a reconciliation loop in restore will be enough`.  Since etcd-operator and restore-operator works together, I think when adding ` reconciliation loop` into restore-operator might requires changes in etcd-operator too. But i am not exactly sure. Let's think about this more and write a design proposal based on it.

**Comment by rjtsdl** [comment]
@fanminshi sounds good. You got what I am trying to do :)

I am working on something for backup, which is more straightforward. And no design change for it.

**Comment by hongchaodeng** [comment]
From the discussion here, sounds like this is just an implementation detail of https://github.com/coreos/etcd-operator/issues/1841 for the periodic backup part. Please converge the discussion there.

For the restore part, can you provide any specific problems or questions?

**Comment by rjtsdl** [comment]
@hongchaodeng , actually, this is a more general proposal. It will set up foundation for #1841. 

For restore, the problems will be:
We need a control loop to monitoring the health of a etcd cluster, rather than just do the restore operation (I think it is currently only instructed by creation of CR). And the restore operator should trigger restore operation without human intervention (**At least for simple scenarios at the beginning**). 

It overlays with @fanminshi 's comments. I am still thinking more about the design proposal.

**Comment by hongchaodeng** [comment]
Let's finish the feature of https://github.com/coreos/etcd-operator/issues/1841 first.
There is no obvious advantage to add control for now.

> We need a control loop to monitoring the health of a etcd cluster, rather than just do the restore operation (I think it is currently only instructed by creation of CR). And the restore operator should trigger restore operation without human intervention 

This is what I'm asking. If it is some new ideas/features you have, please create an new issue to describe any specific use case or problems you have. Don't jump to the solution so quickly. We should discuss before deciding the solution.

**Comment by rjtsdl** [comment]
Fine

I can use #1841 for my current change.

**Comment by drewwells** [comment]
The current disaster design doc implies a restore operator maintains clusters which doesn't appear to be the case any longer. https://github.com/coreos/etcd-operator/blob/master/doc/design/disaster_recovery.md
